File: src/main/java/triangleStructure/Triangle.java
package triangleStructure;

import java.awt.Color;
import java.awt.Polygon;

import global.G;

public class Triangle {

    private static final int SIDES = 3;
    private float[] x = new float[SIDES];
    private float[] y = new float[SIDES];
    private Color color;

    Triangle() {
        for (int i = 0; i < SIDES; i++) {
            x[i] = G.getRandFloat();
            y[i] = G.getRandFloat();
        }
        if (G.getTransparentTriangles()) {
            color = new Color(G.getRandInt(256), G.getRandInt(256), G.getRandInt(256), G.getRandInt(256));
        }
        else {
            color = new Color(G.getRandInt(256), G.getRandInt(256), G.getRandInt(256));
        }
    }

    Triangle(float[] px, float[] py, Color c) {
        x = px;
        y = py;
        color = c;
    }

    Color getColor() {
        return color;
    }

    Polygon getPolygon(int width, int height) {
        width++;
        height++;
        int[] xp = new int[SIDES];
        int[] yp = new int[SIDES];
        for (int i = 0; i < SIDES; i++) {
            xp[i] = (int) (x[i] * width);
            yp[i] = (int) (y[i] * height);
        }
        return new Polygon(xp, yp, SIDES);
    }

    float[] getX() {
        return x.clone();
    }

    float[] getY() {
        return y.clone();
    }

    int[] getColorArray() {
        if (G.getTransparentTriangles()) {
            return new int[] {color.getRed(), color.getGreen(), color.getBlue(), color.getAlpha()};
        }
        else {
            return new int[] {color.getRed(), color.getGreen(), color.getBlue()};
        }
    }

    protected Triangle clone() {
        return new Triangle(getX(), getY(), new Color(color.getRed(), color.getGreen(), color.getBlue(), color.getAlpha()));
    }
}

File: src/main/kotlin/triangleStructure/Triangle.kt
package triangleStructure

import global.Global
import java.awt.Color
import java.awt.Polygon

class Triangle {

    companion object {
        private const val SIDES = 3
    }

    private var x = FloatArray(SIDES)

    private var y = FloatArray(SIDES)

    val color: Color

    internal constructor() {
        for (i in 0 until SIDES) {
            x[i] = Global.randFloat
            y[i] = Global.randFloat
        }
        color = if (Global.transparentTriangles) {
            Color(
                Global.getRandInt(256),
                Global.getRandInt(256),
                Global.getRandInt(256),
                Global.getRandInt(256)
            )
        } else {
            Color(
                Global.getRandInt(256),
                Global.getRandInt(256),
                Global.getRandInt(256)
            )
        }
    }

    internal constructor(px: FloatArray, py: FloatArray, c: Color) {
        x = px
        y = py
        color = c
    }

    fun getPolygon(width: Int, height: Int): Polygon {
        val width1 = width + 1
        val height1 = height + 1

        val xp = IntArray(SIDES)
        val yp = IntArray(SIDES)

        for (i in 0 until SIDES) {
            xp[i] = (x[i] * width1).toInt()
            yp[i] = (y[i] * height1).toInt()
        }

        return Polygon(xp, yp, SIDES)
    }

    fun getX() = x.clone()

    fun getY() = y.clone()

    fun getColorArray() = when {
        Global.transparentTriangles ->
            intArrayOf(
                color.red,
                color.green,
                color.blue,
                color.alpha
            )
        else -> intArrayOf(
            color.red,
            color.green,
            color.blue
        )
    }

    fun clone() = Triangle(
        getX(),
        getY(),
        Color(
            color.red,
            color.green,
            color.blue,
            color.alpha
        )
    )
}